contents/node/nodejs.md

```diff
@@ -0,0 +1,206 @@
+# Introduction to Node.js
+
+Author: Rachael Sim
+
+# Overview
+
+_This chapter assumes that the reader is familiar with JavaScript and asynchronous programming. If you are not familiar with asynchronous programming, a good resource to checkout is [You Don't Know JS guide about asynchronous programming](https://github.com/getify/You-Dont-Know-JS/blob/master/async%20&%20performance/README.md#you-dont-know-js-async--performance) as asynchronous programming is key in Node._
+
+*Node.js* is a JavaScript runtime built on Chrome’s V8 JavaScript engine. Node.js uses an event-driven, non-blocking I/O model that makes it lightweight and efficient.
+
+* **Event-driven**: When an event happens e.g. an I/O event is complete, the event handler/callback function will be enqueued in a queue where it will be subsequently scheduled to run by the [event loop](https://medium.com/the-node-js-collection/what-you-should-know-to-really-understand-the-node-js-event-loop-and-its-metrics-c4907b19da4c).
+* **Non-Blocking**: When file and other I/O operations requests are made in Python or Java, they are blocking -- subsequent requests (in the same thread) have to be enqueued and can only be processed after the current operation completes. Meanwhile, the program remains idle. In contrast, Node.js has an event loop which delegates the actual tasks to other systems (e.g. file system and databases) and it is almost never blocked after delegation. While an I/O operation is incomplete, the event loop can still process subsequent requests.
+
+# Benefits
+
+Why should you use Node?
+
+## Benefit 1: Node.js is fast for I/O intensive tasks
+
+This is due to the non-blocking I/O model.
+
+## Benefit 2: Node.js is single-threaded
+
+This means that we can avoid thread overheads and synchronization problems including deadlocks and race conditions.
+
+## Benefit 3: Use JavaScript for both front and back-end development
+
+Using Node for back-end development makes it possible to share and reuse code between the front-end and back-end. Teams can be more efficient and less reliant on documentation as they can read the code directly. The team can also become more cross-functional as developers can contribute to both development.
+
+## Benefit 4: Node has good support
```

Comment:
You may want to place this after benefit 6, by which time the reader has a better appreciation of npm. In fact you can put the npm-related benefits close to each other and sequence them in a way that later ones build on earlier ones.